### PR TITLE
fix(core): always report the first error on unhandled rejection

### PR DESCRIPTION
**Pull request by 0f-0b**
The root cause of denoland/deno_std#3320, I believe, is that `pending_promise_rejections` is a `HashMap` whose entries are in arbitrary order, and as a result either of the two errors (`AddrInUse` and `TypeError`) may be selected when determining which one to report. I changed the field to a `VecDeque` so that the first error (`AddrInUse` in this case) is always selected.

[Minimal reproduction of the issue addressed by this PR](https://cloudy-ferret-37.deno.dev/). This should always print `error: Uncaught (in promise) "first"` but occasionally prints `error: Uncaught (in promise) "second"`.

If this change requires tests, I would like pointers on how to implement them.

### CODE REVIEW

**Review by bartlomieju — COMMENTED** [review]
Thanks for the PR, could you add a test demonstrating what this change is fixing?

File: core/realm.rs [hunk 1]
```diff
@@ -43,7 +43,7 @@ pub(crate) struct ContextState {
   pub(crate) js_format_exception_cb: Option<Rc<v8::Global<v8::Function>>>,
   pub(crate) js_wasm_streaming_cb: Option<Rc<v8::Global<v8::Function>>>,
   pub(crate) pending_promise_rejections:
-    HashMap<v8::Global<v8::Promise>, v8::Global<v8::Value>>,
+    LinkedHashMap<v8::Global<v8::Promise>, v8::Global<v8::Value>>,
```
Comment (nayeemrmn) [inline core/realm.rs:46]: Does `IndexMap` not work for this?

**Comment by 0f-0b** [comment]
> Could you add a test demonstrating what this change is fixing?

I added a test to `cli/tests/integration/run_tests.rs`. When testing locally, I also found that if there exists a lock file containing npm specifiers, the issue disappears, but the error message does not include `(in promise)`. Why is this?

```shellsession
$ cat deno.lock
{
  "version": "2",
  "remote": {},
  "npm": {
    "specifiers": {
      "@types/node": "@types/node@18.11.18"
    },
    "packages": {
      "@types/node@18.11.18": {
        "integrity": "sha512-DHQpWGjyQKSHj3ebjFI/wRKcqQcdR+MoFBygntYOZytCqNfkd2ZC4ARDJ2DQqhjH5p85Nnd3jhUJIXrszFX/JA==",
        "dependencies": {}
      }
    }
  }
}
$ deno run --lock deno.lock cli/tests/testdata/run/unhandled_rejection_order.ts
error: Uncaught 0
$ deno run cli/tests/testdata/run/unhandled_rejection_order.ts
error: Uncaught (in promise) 26
$ target/debug/deno run cli/tests/testdata/run/unhandled_rejection_order.ts
error: Uncaught (in promise) 0
```

File: core/realm.rs (hunk 1 above)
Comment (0f-0b) [reply core/realm.rs:46]: [Its documentation](https://docs.rs/indexmap/1.9.2/indexmap/map/struct.IndexMap.html#method.shift_remove) states that removing an entry while preserving insertion order takes O(n) time on average.

File: cli/tests/testdata/run/unhandled_rejection_order.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+for (let i = 0; i < 100; i++) {
+  Promise.reject(i);
+}
```
Comment (bartlomieju) [inline cli/tests/testdata/run/unhandled_rejection_order.ts:3]: Nice test, very simply reproduction. I would suggest to turn it into a unit test in `core/runtime.rs` instead

File: core/realm.rs (hunk 1 above)
Comment (nayeemrmn) [reply core/realm.rs]: `indexmap` is already a dep of core, let's use it.

Comment (bartlomieju) [reply core/realm.rs]: I think the current solution is fine, it allows to get rid of the nasty:
```rust
    let key = {
      context_state
        .pending_promise_rejections
        .keys()
        .next()
        .unwrap()
        .clone()
    };
    let handle = context_state
      .pending_promise_rejections
      .remove(&key)
      .unwrap();
```

**Review by bartlomieju — APPROVED** [review]
LGTM, nice fix @0f-0b. Thank you!

**Comment by bartlomieju** [comment]
> > Could you add a test demonstrating what this change is fixing?
> 
> I added a test to `cli/tests/integration/run_tests.rs`. When testing locally, I also found that if there exists a lock file containing npm specifiers, the issue disappears, but the error message does not include `(in promise)`. Why is this?
> 
> ```
> $ cat deno.lock
> {
>   "version": "2",
>   "remote": {},
>   "npm": {
>     "specifiers": {
>       "@types/node": "@types/node@18.11.18"
>     },
>     "packages": {
>       "@types/node@18.11.18": {
>         "integrity": "sha512-DHQpWGjyQKSHj3ebjFI/wRKcqQcdR+MoFBygntYOZytCqNfkd2ZC4ARDJ2DQqhjH5p85Nnd3jhUJIXrszFX/JA==",
>         "dependencies": {}
>       }
>     }
>   }
> }
> $ deno run --lock deno.lock cli/tests/testdata/run/unhandled_rejection_order.ts
> error: Uncaught 0
> $ deno run cli/tests/testdata/run/unhandled_rejection_order.ts
> error: Uncaught (in promise) 26
> $ target/debug/deno run cli/tests/testdata/run/unhandled_rejection_order.ts
> error: Uncaught (in promise) 0
> ```

To be honest, I'm not sure. Could you open a separate issue for this problem?

**Pull request merged**